### PR TITLE
feat!: make nested blocks visible to taint

### PR DESCRIPTION
**Pull request by brandonspark**
## What:
This PR just makes nested blocks visible to the taint engine, notably making them usable as sources or sinks.

## Why:
Sometimes we want to write rules which check if bad values go into nested blocks. This enables that functionality.

For instance, consider the following Terraform code:
```
resource "resource_ty" "resource_name" {
  setting1 = true 

  x { 
    sensitive_setting = var.x
  }
}
```

We would like to be able to write rules which track whether taint goes to the `sensitive_setting`. Unfortunately, it is inside of a `resource`, specifically inside of a `Record`. A `Call` inside of a `Record` is not understood, so this is just killed, and never makes it to the IL, and thus it can't be a source or a sink.

This PR just makes sure it makes it there.

## How:
Just made it so that when in `hcl`, the AST to IL conversion allows nested blocks.

Closes PA-2475

PR checklist:

- [X] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [X] Tests included or PR comment includes a reproducible test plan
- [X] Documentation is up-to-date
- [X] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [X] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: src/analyzing/AST_to_IL.ml
Comment (aryx) [inline src/analyzing/AST_to_IL.ml:887]: Could we do this transformation at parsing time in Parse_hcl_tree_sitter.ml?

Comment (brandonspark) [reply src/analyzing/AST_to_IL.ml:887]: i think in principle, yes, because the main reason why we don't already is because some blocks have more than one "parameter":
```
s "parameter1" "parameter2" {
  <stuff>
}
```
this we can't really nicely translate to a `DefStmt`, because we want the matchability. the blocks I'm concerned with are of the form that they don't take in any extra parameters, however:
```
s {
  <stuff>
}
```

so I feel like this could in fact be done at parse time, by translating it to a `DefStmt` defining a record

it would be a digression from the general case of a block, however. you'd see that some blocks would get translated to `DefStmt` and some would become `ExprStmt (Call ...)`. this lack of homogeneity could cause some manner of confusion. what do you think?

Comment (aryx) [reply src/analyzing/AST_to_IL.ml:887]: Could we transform the s "parameter1" "parameter2" { } in a DefStmt FuncDef?
If it's really defining a function, then let's encode it as a function?
and then if it has no parameter, it's a variable and we can use DefStmt VarDef, looks simple no?

Comment (brandonspark) [reply src/analyzing/AST_to_IL.ml:887]: It looks simple, but unfortunately it's not. I said "parameter", but I didn't mean parameter in the sense of a function, I just meant parameter to a block. An instance of this block may be:

resource "type" "name" {
  arg1 = 1
  arg2 = "hi"
}

where "type" is the type the resource ascribes to, "name" is the name of the resource, and the arguments are the parameters used to configure the resource. It may also look like

settings {
  arg1 = true
}

in the case of a nested block. In neither case is either block truly a "function", and it was my predecessor's choice to keep them as Call's, which is the convention I've kept up with. My question is whether the relative cost of special casing the blocks that look like `settings`, at AST_generic time, is less confusing than special casing it at AST_to_IL time.

Comment (brandonspark) [reply src/analyzing/AST_to_IL.ml:887]: It's also worth noting that these are all Terraform specific notions, and that we claim to support "HCL", which is the more general language surrounding it. If we special case, it will go against the general scheme of HCL.

Comment (brandonspark) [reply src/analyzing/AST_to_IL.ml:887]: @aryx any thoughts?

Comment (aryx) [reply src/analyzing/AST_to_IL.ml:887]: I don't think we care about HCL outside terraform, so if those resources are kinda like functions and variables,
I'd rather transform them in that, so then we can reuse all the IL machinery we have as-is, without having to adapt things later. But I don't understand anything to terraform so ...

Comment (brandonspark) [reply src/analyzing/AST_to_IL.ml:887]: @aryx  I am a little hesitant to do this, because I would prefer that we translate things which are of the same syntactic class (blocks in HCL) to the same kind of construct. I feel like I would forget all the time if some blocks translated to `DefStmt`s, and some translated to `Call`s, nevermind the fact that something like
```
lifecycle {
  settings1 = 2
}
```
is not really "defining" a name of "lifecycle", but it is instead just a parameter that happens to be labelled "block", that contains some settings.

And indeed, none of these are really functions in any sense.

Right now, there is a nice invariant in `Terraform.ml` that any `DefStmt`s are strictly introduced via `transform_ast_for_analyzing`, or otherwise our pre-analysis Terraform-specific processing steps. This is nice because they truly are from `variable` blocks which do define names, so I would prefer to keep them as the only sources of truth for things which define names, because otherwise our logic which is intended to interact strictly with those definitions will also coincide with these blocks here, even though they are not definitions. Does that make sense?

Comment (IagoAbal) [reply src/analyzing/AST_to_IL.ml:887]: Looking at `Parse_hcl_tree_sitter`, I think it would be better to translate this as a `DefStmt` as Yoann suggests. Although I would always translate it into a `VarDef`/`FieldDefColon` like `<block-id> = (<block-params>, <block-body>)`. So e.g. this below could be represented as `x = (["parameter1"; "parameter2"], <stuff>)`:

```hcl
s "parameter1" "parameter2" {
  <stuff>
}
```

Comment (IagoAbal) [reply src/analyzing/AST_to_IL.ml:887]: The thing is that the Terraform translation is arbitrary, we use a `Call` as we could use something else. So it's better to change the `Call` to a `DefStmt` or whatever is needed, rather than keep the `Call` for no particular reason and then add extra code in `AST_to_IL` to cope with that.

**Review by aryx — CHANGES_REQUESTED** [review]
We should put this on hold and try to see if we can generate better generic AST from terraform so that we would not need that.

**Comment by brandonspark** [comment]
@aryx In light of what we discussed at the offsite, might we merge this PR to unblock @LewisArdern from writing Terraform rules?

**Review by aryx — APPROVED** [review]

**Pull request merged**
